### ISSUE TITLE
Investigate torch export/AOT inductor

### ISSUE THREAD

**Issue by ElliottKasoar** [Cambridge-ICCS/FTorch#102]
As of ~PyTorch 2, several discussions seem to suggest that TorchScript is no longer in active development (see [here](https://discuss.pytorch.org/t/pytorch-2-and-the-c-interface/168034/4) and [here](https://discuss.pytorch.org/t/consequences-of-pytorch-2-0-on-libtorch/167576)), although it is unlikely to be deprecated soon either.

Newer forms of saving models appear to use [torch.export](https://pytorch.org/docs/stable/export.html) ([tutorial](https://pytorch.org/tutorials/intermediate/torch_export_tutorial.html)) to save a ".pt2" file - the "PyTorch 2.X way to export PyTorch models into standardized model representations, intended to be run on different (i.e. Python-less) environments".

Development seems to be focused around [TorchInductor](https://dev-discuss.pytorch.org/t/torchinductor-a-pytorch-native-compiler-with-define-by-run-ir-and-symbolic-shapes/747) for model compilation, including a new (PyTorch 2.2) ahead-of-time extension of TorchInductor - [AOTInductor](https://pytorch.org/docs/main/torch.compiler_aot_inductor.html), which is designed for creating artefacts for non-Python environments.

These features are currently prototypes, so the interfaces are likely to change significantly, but it seems likely that support would be useful in the not-too-distant future.

See also: [recent portability discussion](https://github.com/pytorch/pytorch/issues/115965)

**Comment by tztsai** [comment]
`torch.export` saves the PyTorch model as a single computation graph in a standardized representation in the `.pt2` file. This intermediate representation captured by AOT Inductor is compatible to many backends including [Triton](https://github.com/openai/triton) and C++/OpenMP. This new feature is said to be subject to backwards compatibility breaking changes. Although `TorchScript` is no longer in active development, its latest stable version will be supported in the foreseeable future. Therefore maybe we can wait till `TorchInductor` has a stable API and becomes more commonly adopted in place of `TorchScript`.

**Comment by ElliottKasoar** [comment]
Quick update that PyTorch 2.4 supports (in beta) freezing for CPU with the AOTInductor through a `TORCHINDUCTOR_FREEZING` environment variable, if it is eventually supported in some form.

Also, less a FTorch problem but just a another note that [torch.compile](https://pytorch.org/docs/stable/generated/torch.compile.html#torch.compile) (same PT2 stack as `torch.export`) seems to requires 2.4 for Python 3.12 support, if it ever comes up.

This isn't necessarily the best place to share this, but I also recently came across the [PyTorch 2 paper](https://pytorch.org/assets/pytorch2-2.pdf) and [other tutorial material](https://github.com/pytorch/workshops/blob/master/ASPLOS_2024/README.md), which looks like a useful reference for previous and new implementations of PyTorch Graph Capture (`torch.jit`, `TorchDynamo` etc.), portability (`AOTInductor` etc.) and other newer features.

**Comment by jatkinson1000** [comment]
Thanks @fcasson for adding a +1 for this.

It is definitely a priority issue and we are actively applying for funds to support this work (grant application submitted yesterday).
We've recently had some internal discussions to set out the plan for how we'd achieve this and would be keen to work with users to implement/test this.

**Comment by jatkinson1000** [comment]
In reading around this I have the following pieces of documentation that are likely to be important in steering the implementation:

- [General PyTorch docs around AOT and Inductor](https://docs.pytorch.org/docs/stable/user_guide/torch_compiler/export.html)
- [C++ Implementation of Inductor](https://docs.pytorch.org/docs/stable/user_guide/torch_compiler/torch.compiler_aot_inductor.html)
- [Torch export tutorial](https://docs.pytorch.org/tutorials/intermediate/torch_export_tutorial.html)
- [Latest `torch.export` docs](https://docs.pytorch.org/docs/stable/user_guide/torch_compiler/export.html)

**Comment by jatkinson1000** [comment]
Assigning and starting to work on this issue.
I will use the below to document work and findings whilst sandboxing with `torch.export` before opening a PR for subsequent work.

**Comment by jatkinson1000** [comment]
Updating Python simplenet to use `torch.export` observations:

- `Dim` and `dynamic_shapes` are optional and relate to doing dynamic batching.
    - Ignoring for now but will return to them later.
    - The "fix" is to update all model inputs to explicitly specify a batching dimension when we want this.
- `torch.export` files use a `.pt2` extension.
- `torch.export` needs gradient behaviour of models, therefore elements in `inference_mode()` need removing from being hardcoded in models.
- `torch._inductor.aoti_compile_and_package` creates a shared library optimised for this machine/architecture
    - using `inductor_configs` will allow optimisation when compiling for the hardware we run on. We can:
        - Set `inductor_configs = {"max_autotune": False}` to disable hardware optimisations
        - Also set `inductor_configs = {"max_autotune_gemm_backends": "ATEN"}` to use a portable backend
        - Finally, consider importing and recompiling on the host for optimal performance.
- If we want portability between machines use `torch.export.save` and `torch.export.load`
    - the load process raises a userwarning about non-writeable buffers during deserialisation. It seems this is common so we probably want to catch this in the new "pt2te".
- `torch.export.export()` can be used for training with autograd! - [source](https://docs.pytorch.org/docs/stable/user_guide/torch_compiler/export.html#export-ir-training-vs-inference)

**Comment by jatkinson1000** [comment]
It seems there is not yet a C++ runtime/interface to `export`.
[Not an open-source one at least...](https://github.com/pytorch/pytorch/issues/103841)

There is an [issue that was asking about the ability to load directly from saved files](https://github.com/pytorch/pytorch/issues/144663). It indicated this was targeted for later 2025 so I have added a bump to request an update.
